File: nodejs/src/models/Order.ts
import { Beverage } from "../decorator/BeverageDecorator";
import OrderState, { IOrder } from "../state/OrderState";
import NewOrderState from "../state/NewOrderState";
import { PaymentStrategy } from "../strategy/PaymentStrategy";

class Order implements IOrder {
    id: string;
    items: Beverage[];
    private state: OrderState;
    private paymentStrategy: PaymentStrategy | null;

    constructor(id: string) {
        this.id = id;
        this.items = [];
        this.paymentStrategy = null;
        this.state = new NewOrderState(this);
    }

    setState(state: OrderState): void {
        this.state = state;
    }

    getStateName(): string {
        return this.state.getName();
    }

    next(): void {
        this.state.next();
    }

    cancel(): void {
        this.state.cancel();
    }

    addItem(beverage: Beverage): void {
        if (this.state.addItem(beverage)) {
            this.items.push(beverage);
        }
    }

    getTotal(): number {
        return this.items.reduce((sum, item) => sum + item.getCost(), 0);
    }

    setPaymentStrategy(strategy: PaymentStrategy): void {
        this.paymentStrategy = strategy;
    }

    pay(): void {
        if (!this.paymentStrategy) {
            console.log("⚠️ Chưa chọn phương thức thanh toán");
            return;
        }
        const result = this.paymentStrategy.pay(this.getTotal());
        console.log(`[${this.paymentStrategy.getName()}] ${result.message}`);
    }
}

export default Order;

File: spring/demo/src/main/java/com/example/demo/model/Order.java
package com.example.demo.model;

import com.example.demo.decorator.Beverage;
import com.example.demo.state.NewOrderState;
import com.example.demo.state.OrderState;
import com.example.demo.strategy.PaymentResult;
import com.example.demo.strategy.PaymentStrategy;

import java.util.ArrayList;
import java.util.List;

public class Order {
    private final String id;
    private final List<Beverage> items = new ArrayList<>();
    private OrderState state;
    private PaymentStrategy paymentStrategy;

    public Order(String id) {
        this.id = id;
        this.state = new NewOrderState(this);
    }

    public String getId() {
        return id;
    }

    public void setState(OrderState state) {
        this.state = state;
    }

    public String getStateName() {
        return state.getName();
    }

    public void next() {
        state.next();
    }

    public void cancel() {
        state.cancel();
    }

    public void addItem(Beverage beverage) {
        if (state.addItem(beverage)) {
            items.add(beverage);
        }
    }

    public int getTotal() {
        return items.stream().mapToInt(Beverage::getCost).sum();
    }

    public void setPaymentStrategy(PaymentStrategy strategy) {
        this.paymentStrategy = strategy;
    }

    public PaymentResult pay() {
        if (paymentStrategy == null) {
            System.out.println("Chưa chọn phương thức thanh toán");
            return null;
        }

        PaymentResult result = paymentStrategy.pay(getTotal());
        System.out.println("[" + paymentStrategy.getName() + "] " + result.getMessage());
        return result;
    }
}
